lib/AST/ClangTypeConverter.cpp

```diff
@@ -883,6 +900,93 @@ Decl *ClangTypeConverter::getSwiftDeclForExportedClangDecl(
   return (it != ReversedExportMap.end() ? it->second : nullptr);
 }
 
+clang::QualType ClangTypeConverter::convertTemplateArgument(Type type) {
+  // C++ function templates can only be instantiated with Clang types and
+  // a handful of Swift builtin types. These are enumerated here rather than
+  // delegated to ClangTypeConverter::convert() (which is more general).
+  auto withCache = [&](auto lookup) {
+    auto cached = Cache.find(type);
+    if (cached != Cache.end())
+      return cached->second;
+
+    auto result = lookup();
+    if (!result.isNull())
+      Cache.insert({type, result});
+    return result;
+  };
+
+  // This type was imported from Clang, so we can convert it back by retrieving
+  // ClangDecl stored in the imported type decl (without making a cache entry.)
+  if (auto nominal = type->getAs<NominalType>())
```

Comment:
I am a bit on the edge whether we want to use a `TypeVisitor` here, but maybe it is OK as is for now.